consensus/istanbul/core/commit.go

```diff
@@ -177,10 +189,9 @@ func (c *core) handleCheckedCommitForPreviousSequence(msg *istanbul.Message, com
 	if validator == nil {
 		return errInvalidValidatorAddress
 	}
-	if err := c.verifyCommittedSeal(commit, validator); err != nil {
-		return errInvalidCommittedSeal
-	}
```

Comment:
I didn't see anywhere in the code that a bad signature introduced the parent seal at this point would be handled. I'd be worried that this means a validator can compromise liveness by sending a bad commit for the previous sequence every round.

Reply:
This is handled here https://github.com/celo-org/celo-blockchain/blob/d759429e620d0098f76f35ef73648dca6c2a6df3/consensus/istanbul/backend/engine.go#L983-L1024

The process to build the parent seal is to take the previous block seal as a starting point and combine any new commit messages encountered while processing the current block, if the aggregate seal does not verify then the previous block seal is used. So if there is one bad signature in any late arriving commits no late arriving commits will be added to the parent seal.

Reply:
hmm, this is technically correct behavior, but it makes it possible for a single bad validator to "mark" the "late" (actually anything past quorum) validators as down for the block. There's definitely overhead to it, but I think it makes sense to do the check again on the late arriving commits.

Reply:
This is now done